### DISCUSSION TITLE
Instructions for compiling against Cosmopolitan libc?

### DISCUSSION THREAD

**Discussion by neezer** [helix-editor/helix#5387] [Ideas]
👋 I'm trying out OpenBSD (and loving it!) but 7.2 stable doesn't include a package for Helix 😢 . [Snapshots do](https://openports.se/editors/helix), but I've got other issues with`-current` that will likely prevent me from using that.

Soooo, in the meantime, I'm trying to figure out ways to get Helix on my OpenBSD system (among many other Rust programs I want). `cargo install` doesn't work for many Rust packages right now because of [an issue with system curl](https://github.com/rust-lang/cargo/issues/11435) (I think). I tried building a statically-linked binary on macOS per [this article](https://msfjarvis.dev/posts/building-static-rust-binaries-for-linux/) and copying the artifacts over to OpenBSD but haven't had luck there either.

Then I found this [delightfully tantalizing approach](https://ahgamut.github.io/2022/07/27/ape-rust-example/), which uses [Cosmopolitan libc](https://justine.lol/cosmopolitan/) as a custom cargo build target. I'm not an expert in C but it appears as if this would allow me to compile a portable executable on one machine and copy it over (with Helix-specific supporting files) to OpenBSD and have it Just Work™.

The author of the last article managed to do this with [ripgrep](https://github.com/BurntSushi/ripgrep) ([example repo compiled against Cosmopolitan libc](https://github.com/ahgamut/ripgrep/tree/cosmopolitan)), which seemed to me roughly equivalent in complexity to Helix, further fueling my notion that this could actually work. I've tinkered with cloning the ripgrep example and tweaking for Helix and attempting to compile on a Debian system, but I haven't gotten this to build correctly yet. From my own experience--and from the article's notes--I think most of the complexity is figuring out how to tell cargo to "just use cosmopolitan.a for everything":

> I’m pretty sure a lot of software can be built with Cosmopolitan Libc, it’s just a matter of convincing the build system.

So, I figured I'd put the question to y'all, who likely know far more than I do about whether or not an approach like this would be possible with Helix. If it is, I feel like this might be a great strategy for distribution, as one binary would work on "Linux + Mac + Windows + FreeBSD + OpenBSD + NetBSD + BIOS," according to the author of Cosmopolitan libc.

If Cosmo is too young and unproven to consider using for official distributions, I'd still love instructions for the adventurous like me to compile Helix against it while we wait for it to mature.

**Comment by NoelJacob** [comment]
Of the top of my head try compiling to WASM then WASM to c and finally cosmocc
